teps/0002-custom-tasks.md

```diff
@@ -0,0 +1,392 @@
+
+Enable integrators to define new "custom" Task types as CRDs, which can be run
+by creating a new `Run` object (a CRD type which Tekton will define and
+own). Custom Task authors must also implement a reconciling controller which
+watches for creations of `Run` objects which reference their type, and
+which eventually updates its status to signal task completion.
+
+To achieve this, Tekton will define a new type, `Run`, first in
+`v1alpha1`, with the intention of iterating and, if there's support, bringing
+this functionality to `v1beta1` and eventually `v1`. Tekton will also
+implement support for this new type in the PipelineRun CRD controller, as well
+as document the contract for integrating with this new type, and providing
+examples and Go helper methods that integrators can consume if their
+controllers are implemented in Go.
+
+## Motivation
+
+Allow integrators to implement new functionality that doesn't directly map to
+Tekton's current idea of a `Task` and `TaskRun` (i.e., a collection of containers
+running in a `Pod` on a compute node).
+
+Some examples include (non-exhaustive):
+
+* wait for a period of time without having to schedule a container that just sleeps, wasting compute resources
+* wait for an external event to occur, e.g., an approval event signal
+* execute some operation outside of the cluster (e.g., a cloud build service, a macOS/mobile build farm) and wait for its execution to complete
+* execute another (sub-)Pipeline and wait for it to complete
+* enable matrix parameter expression -- e.g., succinctly express a Task that executes with args A,B,C (defined at config-time)
+* enable looping execution of sub-Tasks -- e.g., express that a Task should be repeatedly run until some state is reached (signalled at run-time)
+* ...and in general, support a model where integrators can implement their own execution types without modifying Tekton Pipelines code directly
+
+This mechanism can also be used by Tekton core contributors to prototype and experiment with new execution modes, even other forms of Pod-based Tasks, before possibly integrating that functionality into Tekton Pipelines directly.
+
+### Goals
+
+1. Allow non-Pod Task implementations to be built and integrated into "core"
+Tekton Pipelines.
+
+1. Implement a handful of commonly-requested features (e.g., wait, approval)
+as example Custom Tasks in
+[tektoncd/experimental](https://github.com/tektoncd/experimental), to
+demonstrate the contract and act as a working example for future integrators.
+
+1. Provide Go packages and frameworks to help integrators perform common operations required by the Custom Task implementation contract (described below).
+
+### Non-Goals
+
+1. Provide any Custom Task implementations as "official" or "first-party"
+integrations. Some may be added in a future change, but for now the goal
+is just to support _any_ third-party integrations, and let the user install them themselves, or let distributors provide them if they prefer to.
```

Comment:
Are you saying that you don't want "official" custom tasks built on top of this support in the near future by anybody, or just in the scope of this TEP?  I ask because I've been looking at implementing task looping using custom tasks and @jerop is looking at implementing conditions using custom tasks.  I would expect those items be assessed by the community on their own as to whether they should be official or experimental -- unless you are stating a position that it's premature to implement any custom tasks officially?

Reply:
@GregDritschler what @vdemeester said. This TEP should only cover _being able to write a custom task_, and not expand to include _any official custom tasks_.

I can definitely see a future where the `tektoncd/pipeline` repo includes some custom task controllers for commonly requested features, maintained and supported by Tekton contributors. But, in the future a Tekton platform/distribution (e.g., Openshift Pipelines) might decide not to install some that they don't need, or install their own versions, or install new ones by default, etc. -- that's up to them.